src/jit/gentree.cpp

```diff
@@ -9183,259 +9183,344 @@ GenTreePtr GenTree::GetChild(unsigned childNum)
-            {
-                return &m_node->AsStmt()->gtStmtExpr;
-            }
-            return nullptr;
+            case GT_FIELD:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceField;
+                break;
 
-        case GT_ARR_ELEM:
-            if (m_state == 0)
-            {
-                return &m_node->AsArrElem()->gtArrObj;
-            }
-            else if (m_state <= m_node->AsArrElem()->gtArrRank)
-            {
-                return &m_node->AsArrElem()->gtArrInds[m_state - 1];
-            }
-            return nullptr;
+            case GT_STMT:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStmt;
+                break;
 
-        case GT_ARR_OFFSET:
-            switch (m_state)
-            {
-                case 0:
-                    return &m_node->AsArrOffs()->gtOffset;
-                case 1:
-                    return &m_node->AsArrOffs()->gtIndex;
-                case 2:
-                    return &m_node->AsArrOffs()->gtArrObj;
-                default:
-                    return nullptr;
-            }
+            case GT_ARR_ELEM:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrElem;
+                break;
 
-        // Call, phi, and SIMD nodes are handled by MoveNext{Call,Phi,SIMD}UseEdge, repsectively.
-        case GT_CALL:
-        case GT_PHI:
-#ifdef FEATURE_SIMD
-        case GT_SIMD:
-#endif
-            break;
+            case GT_ARR_OFFSET:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceArrOffset;
+                break;
+
+            case GT_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceDynBlk;
+                break;
 
-        case GT_ASG:
+            case GT_STORE_DYN_BLK:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceStoreDynBlk;
+                break;
+
+            case GT_CALL:
+                m_advance = &GenTreeUseEdgeIterator::AdvanceCall;
+                SetEntryStateForCall();
+                return;
+
+            default:
+                unreached();
+        }
+    }
+
+    // Advance to the first operand.
+    ++(*this);
+}
+
+void GenTreeUseEdgeIterator::AdvanceCmpXchg()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
+        {
+            case 0:
+                return &m_node->AsCmpXchg()->gtOpLocation;
+            case 1:
+                return &m_node->AsCmpXchg()->gtOpValue;
+            case 2:
+                return &m_node->AsCmpXchg()->gtOpComparand;
+            default:
+                return nullptr;
+        }
+    }());
+}
+
+void GenTreeUseEdgeIterator::AdvanceBoundsChk()
+{
+    FinishSimpleAdvance([this]() -> GenTree** {
+        switch (m_state)
+        {
+            case 0:
+                return &m_node->AsBoundsChk()->gtIndex;
+            case 1:
+                return &m_node->AsBoundsChk()->gtArrLen;
+            default:
+                return nullptr;
+        }
+    }());
+}
```

Comment:
FinishSimpleAdvance is checking `if (m_edge != nullptr && *m_edge != nullptr)`. That is unnecessary in many cases: when we've returned the `&m_node...` we know `m_edge != nullptr`, and we know from the IR that many operands are not allowed to be nullptr, so the `*m_edge != nullptr`, e.g., see NullOp1Legal()/NullOp2Legal(). In many cases you can just assert on the non-null-ness and simply do `m_state++`

---
In reply to: [105434088](https://github.com/dotnet/coreclr/pull/10054#discussion_r105434088) [](ancestors = 105434088)